From 394b5132551b269f8804940e634a14095d82a28c Mon Sep 17 00:00:00 2001
From: Shinobu Uehara <shinobu.uehara.xc@renesas.com>
Date: Fri, 14 Jun 2013 12:03:38 +0900
Subject: [PATCH 0509/1083] mmc: tmio: Add CTL_RESET_SDIO register access pass
 option

Signed-off-by: Shinobu Uehara <shinobu.uehara.xc@renesas.com>
(cherry picked from commit 413bd064b79f237c9bfff40eb51fbf1d209154ef)

Signed-off-by: Ryo Kataoka <ryo.kataoka.wt@renesas.com>
---
 drivers/mmc/host/tmio_mmc_pio.c | 8 ++++++--
 include/linux/mfd/tmio.h        | 3 +++
 2 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/host/tmio_mmc_pio.c b/drivers/mmc/host/tmio_mmc_pio.c
index 0d8a9bb..0cf4a3a 100644
--- a/drivers/mmc/host/tmio_mmc_pio.c
+++ b/drivers/mmc/host/tmio_mmc_pio.c
@@ -1,6 +1,7 @@
 /*
  * linux/drivers/mmc/host/tmio_mmc_pio.c
  *
+ * Copyright (C) 2013 Renesas Electronics Corporation
  * Copyright (C) 2011 Guennadi Liakhovetski
  * Copyright (C) 2007 Ian Molton
  * Copyright (C) 2004 Ian Molton
@@ -190,15 +191,18 @@ static void tmio_mmc_clk_start(struct tmio_mmc_host *host)
 static void tmio_mmc_reset(struct tmio_mmc_host *host)
 {
 	struct resource *res = platform_get_resource(host->pdev, IORESOURCE_MEM, 0);
+	struct tmio_mmc_data *pdata = host->pdata;
 
 	/* FIXME - should we set stop clock reg here */
 	sd_ctrl_write16(host, CTL_RESET_SD, 0x0000);
 	/* implicit BUG_ON(!res) */
-	if (resource_size(res) > 0x100)
+	if (!(pdata->flags & TMIO_MMC_NO_CTL_RESET_SDIO)
+		&& resource_size(res) > 0x100)
 		sd_ctrl_write16(host, CTL_RESET_SDIO, 0x0000);
 	msleep(10);
 	sd_ctrl_write16(host, CTL_RESET_SD, 0x0001);
-	if (resource_size(res) > 0x100)
+	if (!(pdata->flags & TMIO_MMC_NO_CTL_RESET_SDIO)
+		&& resource_size(res) > 0x100)
 		sd_ctrl_write16(host, CTL_RESET_SDIO, 0x0001);
 	msleep(10);
 }
diff --git a/include/linux/mfd/tmio.h b/include/linux/mfd/tmio.h
index 9d35929..c19c5fc 100644
--- a/include/linux/mfd/tmio.h
+++ b/include/linux/mfd/tmio.h
@@ -85,6 +85,9 @@
 /* Switch the register of controller to access the buffer of 16 bit. */
 #define TMIO_MMC_BUFF_16BITACC_ACTIVE_HIGH	(1 << 6)
 
+/* NO_CTL_RESET_SDIO register don't work. */
+#define TMIO_MMC_NO_CTL_RESET_SDIO	(1 << 7)
+
 int tmio_core_mmc_enable(void __iomem *cnf, int shift, unsigned long base);
 int tmio_core_mmc_resume(void __iomem *cnf, int shift, unsigned long base);
 void tmio_core_mmc_pwr(void __iomem *cnf, int shift, int state);
-- 
1.8.3.2

